### PR TITLE
Spark-Enabled Viewshed

### PR DESCRIPTION
**Pull request by jamesmcclain**
Contains
   - Single-tile implementation of the R2 viewshed algorithm from [1]
      - [x] Basics
      - [x] Account for curvature of Earth
      - [x] Choice of AND, OR, DEBUG aggregation operators
   - Distributed viewshed using R2
      - [x] Singe point case
      - [x] Multiple point case
   - Validation
      - [x] Use cases
   - [x] Extension methods

1. Franklin, Wm Randolph, and Clark Ray.
   "Higher isn’t necessarily better: Visibility algorithms and experiments."
    Advances in GIS research: sixth international symposium on spatial data handling. Vol. 2.
    Taylor & Francis Edinburgh, 1994.

Fixes locationtech/geotrellis#1981

-------

Points at (-10877415, 4448309) and (-10940657, 4501659), respectively.  Both are 4 km above the surface.

screenshot from 2017-03-08 21 08 57

screenshot from 2017-03-08 21 09 10

-------

Field of view: one 45 degrees and the other 315 degrees.

screenshot from 2017-03-13 13 12 50

-------

Using the altitude parameter: the 315 degree wedge queries above the surface.

screenshot from 2017-03-13 16 32 06

### CODE REVIEW

File: spark/src/main/scala/geotrellis/spark/viewshed/IterativeViewshed.scala [hunk 1]
```diff
@@ -0,0 +1,390 @@
...
+  *    Advances in GIS research: sixth international symposium on spatial data handling. Vol. 2.
+  *    Taylor & Francis Edinburgh, 1994.
+  */
+object IterativeViewshed {
+
+  /**
+    * A `Point6D` is an array of six doubles.  The indices are mapped as follows:
+    *   0: x-coordinate (in the units used by the layer)
+    *   1: y-coordinate (in the units used by the layer)
+    *   2: view height (in units of "meters")
```
Comment (echeipesh) [inline spark/src/main/scala/geotrellis/spark/viewshed/IterativeViewshed.scala]: Looking at the algorithm this is height from the surface at the given point. Is that what we expect the definition to be or is it expected to be absolute height? + @lossyrob @rossbernet

File: spark/src/main/scala/geotrellis/spark/viewshed/IterativeViewshed.scala [hunk 2]
```diff
@@ -0,0 +1,390 @@
...
+    val _pointsByKey: Map[SpatialKey, Seq[PointInfo]] =
+      info
+        .groupBy(_.key)
+        .toMap
+    val pointsByKey = sc.broadcast(_pointsByKey)
+    if (touchedKeys != null) touchedKeys ++= _pointsByKey.keys
+
+    val _pointsByIndex: Map[Int, PointInfo] =
+      info
+        .groupBy(_.index)
```
Comment (echeipesh) [inline spark/src/main/scala/geotrellis/spark/viewshed/IterativeViewshed.scala]: Checking for my understanding, the `info` was created from `zipWithIndex` why not just take it in the order that its created?

File: spark/src/main/scala/geotrellis/spark/viewshed/IterativeViewshed.scala [hunk 3]
```diff
@@ -0,0 +1,390 @@
...
+    // Repeatedly  map over the RDD  of viewshed tiles until  all rays
+    // have reached the periphery of the layer.
+    do {
+      val _changes: Map[SpatialKey, Seq[(Int, From, mutable.ArrayBuffer[Ray])]] =
+        rays.value
+          .groupBy(_._1)
+          .map({ case (k, list) =>
+            (k, list.map({ case (_, index, from, rs) => (index, from, rs) }))
+          })
+          .toMap
```
Comment (echeipesh) [inline spark/src/main/scala/geotrellis/spark/viewshed/IterativeViewshed.scala]: How big are these change sets typically? It seems like this is this sorting, and the sort by index few lines below, can happen in the accumulator easily enough but not clear if the optimization is likely to be material.

**Comment by pomadchin** [comment]
WARN: this PR depends on https://github.com/locationtech/geotrellis/pull/2226 which still requires CQs

File: spark/src/main/scala/geotrellis/spark/viewshed/IterativeViewshed.scala (hunk 1 above)
Comment (jamesmcclain) [inline spark/src/main/scala/geotrellis/spark/viewshed/IterativeViewshed.scala]: I don't know what sociological crust may exist around this, but this is sort of required to be logically consistent.

File: raster/src/main/scala/geotrellis/raster/viewshed/R2Viewshed.scala [hunk 4]
```diff
@@ -0,0 +1,440 @@
...
+  *
+  * 1. Franklin, Wm Randolph, and Clark Ray.
+  *    "Higher isn’t necessarily better: Visibility algorithms and experiments."
+  *    Advances in GIS research: sixth international symposium on spatial data handling. Vol. 2.
+  *    Taylor & Francis Edinburgh, 1994.
+  */
+object R2Viewshed extends Serializable {
+
+  sealed abstract class From()
+  case class FromNorth() extends From
```
Comment (jamesmcclain) [inline raster/src/main/scala/geotrellis/raster/viewshed/R2Viewshed.scala]: Fixed

File: raster/src/main/scala/geotrellis/raster/viewshed/R2Viewshed.scala [hunk 5]
```diff
@@ -0,0 +1,440 @@
...
+  case class FromEast() extends From
+  case class FromSouth() extends From
+  case class FromWest() extends From
+  case class FromInside() extends From
+
+  sealed abstract class AggregationOperator()
+  case class And() extends AggregationOperator { override def toString: String = "AND" }
```
Comment (jamesmcclain) [inline raster/src/main/scala/geotrellis/raster/viewshed/R2Viewshed.scala]: Fixed

File: spark/src/main/scala/geotrellis/spark/viewshed/IterativeViewshed.scala [hunk 6]
```diff
@@ -0,0 +1,390 @@
...
+  /**
+    * A `Point6D` is an array of six doubles.  The indices are mapped as follows:
+    *   0: x-coordinate (in the units used by the layer)
+    *   1: y-coordinate (in the units used by the layer)
+    *   2: view height (in units of "meters")
+    *   3: the angle in radians (about the z-axis) of the "camera"
+    *   4: the field of view of the "camera" in radians
+    *   5: the absolute altitude to query; if -∞ then use the terrain height
+    */
+  type Point6D = Array[Double]
```
Comment (jamesmcclain) [inline spark/src/main/scala/geotrellis/spark/viewshed/IterativeViewshed.scala]: Fixed

File: spark/src/main/scala/geotrellis/spark/viewshed/IterativeViewshed.scala [hunk 7]
```diff
@@ -0,0 +1,390 @@
...
+
+  implicit def coordinatesToPoints(points: Seq[jts.Coordinate]): Seq[Point6D] =
+    points.map({ p => Array(p.x, p.y, p.z, 0, -1.0, Double.NegativeInfinity) })
+
+  private val logger = Logger.getLogger(IterativeViewshed.getClass)
+
+  private type Message = (SpatialKey, Int, From, mutable.ArrayBuffer[Ray]) // key, point index, direction, rays
```
Comment (jamesmcclain) [inline spark/src/main/scala/geotrellis/spark/viewshed/IterativeViewshed.scala]: Fixed

File: spark/src/main/scala/geotrellis/spark/viewshed/IterativeViewshed.scala [hunk 8]
```diff
@@ -0,0 +1,390 @@
...
+
+  /**
+    * Compute the resolution (in meters per pixel) of a layer.
+    */
+  private def computeResolution[K: (? => SpatialKey): ClassTag, V: (? => Tile)](
+    elevation: RDD[(K, V)] with Metadata[TileLayerMetadata[K]]
+  ) = {
+    val md = elevation.metadata
+    val mt = md.mapTransform
+    val key = implicitly[SpatialKey](elevation.map(_._1).first)
```
Comment (jamesmcclain) [inline spark/src/main/scala/geotrellis/spark/viewshed/IterativeViewshed.scala]: Fixed

File: spark/src/main/scala/geotrellis/spark/viewshed/IterativeViewshed.scala [hunk 9]
```diff
@@ -0,0 +1,390 @@
...
+    val md = rdd.metadata
+
+    val p2 = new jts.Coordinate(p(0),p(1))
+    val bounds = md.layout.mapTransform(p2.envelope)
+    require(bounds.colMin == bounds.colMax)
+    require(bounds.rowMin == bounds.rowMax)
+
+    val cols = md.layout.tileCols
+    val rows = md.layout.tileRows
+    val key = SpatialKey(bounds.colMin, bounds.rowMin)
```
Comment (jamesmcclain) [inline spark/src/main/scala/geotrellis/spark/viewshed/IterativeViewshed.scala]: I am afraid that I do not follow.

File: spark/src/main/scala/geotrellis/spark/viewshed/IterativeViewshed.scala [hunk 10]
```diff
@@ -0,0 +1,390 @@
...
+  }
+
+  /**
+    * The main entry-point for the iterative viewshed implementation.
+    * Takes a layer, some source points, and other ancillary
+    * information and produces a viewshed layer.
+    *
+    * @param  elevation    The elevation layer; pixel values are interpreted as being in units of "meters"
+    * @param  ps           Viewshed source points; their construction and interpretation is described above
+    * @param  maxDistance  The maximum distance that rays are allowed to travel; a lower number reduces computational cost
```
Comment (jamesmcclain) [inline spark/src/main/scala/geotrellis/spark/viewshed/IterativeViewshed.scala]: Fixed

File: spark/src/main/scala/geotrellis/spark/viewshed/IterativeViewshed.scala [hunk 11]
```diff
@@ -0,0 +1,390 @@
...
+    * @param  curvature    Whether or not to take the curvature of the Earth into account
+    * @param  operator     The aggregation operator to use (e.g. Or)
+    * @param  epsilon      Rays within this many radians of horizontal (vertical) are considered to be horizontal (vertical)
+    * @param  touchedKeys  A set of SpatialKeys to keep track of which keys have been touched
```
Comment (jamesmcclain) [inline spark/src/main/scala/geotrellis/spark/viewshed/IterativeViewshed.scala]: Fixed

File: spark/src/main/scala/geotrellis/spark/viewshed/IterativeViewshed.scala [hunk 12]
```diff
@@ -0,0 +1,390 @@
...
+    */
+  def apply[K: (? => SpatialKey): ClassTag, V: (? => Tile)](
+    elevation: RDD[(K, V)] with Metadata[TileLayerMetadata[K]],
+    ps: Seq[Point6D],
+    maxDistance: Double,
+    curvature: Boolean = true,
+    operator: AggregationOperator = Or(),
+    epsilon: Double = (1/math.Pi),
+    touchedKeys: mutable.Set[SpatialKey] = null
+  )(implicit sc: SparkContext): RDD[(K, Tile)] with Metadata[TileLayerMetadata[K]] = {
```
Comment (jamesmcclain) [inline spark/src/main/scala/geotrellis/spark/viewshed/IterativeViewshed.scala]: Fixed

File: spark/src/main/scala/geotrellis/spark/viewshed/IterativeViewshed.scala [hunk 13]
```diff
@@ -0,0 +1,390 @@
...
+
+    val md = elevation.metadata
+    val mt = md.mapTransform
+
+    val resolution = computeResolution(elevation)
+    logger.debug(s"Computed resolution: $resolution meters/pixel")
+
+    val bounds = md.bounds match {
+      case b: KeyBounds[K] => b
+      case _ => throw new Exception
```
Comment (jamesmcclain) [inline spark/src/main/scala/geotrellis/spark/viewshed/IterativeViewshed.scala]: Fixed

File: spark/src/main/scala/geotrellis/spark/viewshed/IterativeViewshed.scala [hunk 14]
```diff
@@ -0,0 +1,390 @@
...
+      * to interested neighbors.
+      */
+    def rayCatcherFn(key: SpatialKey, index: Int)(bundle: Bundle): Unit = {
+      val southKey = SpatialKey(key.col + 0, key.row + 1)
+      val westKey = SpatialKey(key.col + 1, key.row + 0)
+      val northKey = SpatialKey(key.col + 0, key.row - 1)
+      val eastKey = SpatialKey(key.col - 1, key.row + 0)
+
+      Map(
+        FromSouth() -> southKey,
```
Comment (jamesmcclain) [inline spark/src/main/scala/geotrellis/spark/viewshed/IterativeViewshed.scala]: Fixed

File: spark/src/main/scala/geotrellis/spark/viewshed/IterativeViewshed.scala (hunk 2 above)
Comment (jamesmcclain) [inline spark/src/main/scala/geotrellis/spark/viewshed/IterativeViewshed.scala]: Fixed

File: spark/src/main/scala/geotrellis/spark/viewshed/IterativeViewshed.scala (hunk 3 above)
Comment (jamesmcclain) [inline spark/src/main/scala/geotrellis/spark/viewshed/IterativeViewshed.scala]: That is a good point;  these sets are in general bounded below in size by the square root of the number of tiles times the number of points.

**Comment by jamesmcclain** [comment]
> WARN: this PR depends on #2226 which still requires CQs

PR no longer relies on #2226

File: spark/src/main/scala/geotrellis/spark/viewshed/IterativeViewshed.scala (hunk 3 above)
Comment (jamesmcclain) [inline spark/src/main/scala/geotrellis/spark/viewshed/IterativeViewshed.scala]: Fixed

**Comment by jamesmcclain** [comment]
All comments addressed.

**Pull request merged**